stew/bitops2.nim

```diff
@@ -394,3 +401,147 @@ func rotateRight*(v: SomeUnsignedInt, amount: SomeInteger):
   const mask = bitsof(v) - 1
   let amount = int(amount and mask)
   (v shr amount) or (v shl ( (-amount) and mask))
+
+template mostSignificantBit(T: type): auto =
+  const res = 1 shl (sizeof(T) * 8 - 1)
+  T(res)
+
+template getBit*(x: BitIndexable, bit: Natural): bool =
+  ## reads a bit from `x`, assuming 0 to be the position of the
+  ## least significant bit
+  type T = type(x)
+  (x and T(0b1 shl bit)) != 0
+
+template getBitLE*(x: BitIndexable, bit: Natural): bool =
+  getBit(x, bit)
+
+template getBitBE*(x: BitIndexable, bit: Natural): bool =
+  ## Reads a bit from `x`, assuming 0 to be the position of
+  ## the most significant bit.
+  ##
+  ## This indexing may be natural when you are considering the
+  ## string representation of a bit field. For example, 72 can
+  ## be written in binary as 0b01001000. The first bit here is
+  ## zero, while the second bit is one.
+  ##
+  ## Since the string representation will depend on the size of
+  ## the operand, using `getBitBE` with the same numeric value
+  ## and a bit position may produce different results depending
+  ## on the machine type used to store the value. For this reason,
+  ## this indexing scheme is considered more error-prone and
+  ## `getBitLE` is considering the default indexing scheme.
+  (x and mostSignificantBit(x.type) shr bit) != 0
+
+proc setBit*(x: var BitIndexable, bit: Natural, val: bool) =
+  ## writes a bit in `x`, assuming 0 to be the position of the
+  ## least significant bit
+  type T = type(x)
+  let mask = T(0b1 shl bit)
+  if val:
+    x = x or mask
+  else:
+    x = x and not mask
+
+proc setBitLE*(x: var BitIndexable, bit: Natural, val: bool) =
+  setBit(x, bit, val)
+
+proc setBitBE*(x: var BitIndexable, bit: Natural, val: bool) =
+  ## writes a bit in `x`, assuming 0 to be the position of the
+  ## most significant bit
+  let mask = mostSignificantBit(x.type) shr bit
+  if val:
+    x = x or mask
+  else:
+    x = x and not mask
+
+proc raiseBit*(x: var BitIndexable, bit: Natural) =
```

Comment:
https://gcc.godbolt.org/z/cXfo2O

Reply:
I've found `raiseBit` and `lowerBit` to be quite popular in our code in practice. I don't see why we should make the life of the compiler any harder. What do we win exactly?

Reply:
fewer names to remember, more alignment with other languages? `raise` and `lower` don't look like "established" names for this stuff so there's a hurdle to understand what they do (https://www.google.com/search?client=firefox-b-d&ei=S0dBXczOLciAgQa7-pOACw&q=raise+lower+bit&oq=raise+lower+bit), `set` and `clear` I'd say are generally used/accepted

Reply:
Since bits are sometimes called "flags", I've seen "raise" being used for them, because a real flag can be raised and lowered. `set` is more ambiguous for me, because it also used for setters.

Reply:
in that case we should use the same terminology in `bitops`, talking about raised bits instead of set ones.

that said, `set` and `clear` are really established terms for this - used all over the  place from [bit hack docs](http://graphics.stanford.edu/~seander/bithacks.html), [atomic flags](https://en.cppreference.com/w/cpp/atomic/atomic_flag), [stackoverflow](https://stackoverflow.com/questions/47981/how-do-you-set-clear-and-toggle-a-single-bit).

`raise` on the other hand is used in the language in conjunction with exceptions already. 

it certainly adds to nim's uniqueness to invent new terms for these low level ops, but I don't think in a meaningful / good way (akin to using your own style in a small part of a larger code base)

Reply:
Well, I appreciate this attention to detail and I encourage everyone to carry out modifications like this. But I do have a long list of probably more important tasks that I can work on in the near term.

Reply:
And to clarify, I still disagree that the "optimized" short-cuts such as `raiseBit` are unnecessary.

Reply:
no need to block the PR for this stuff, but it's good to establish what we want, in this matter - refactoring can be done at a later stage, it's kind of trivial.

Reply:
The gold standard of bit manipulation is "Bit Twiddling Hack" and they use set/clear - https://graphics.stanford.edu/~seander/bithacks.html#ConditionalSetOrClearBitsWithoutBranching

`test` is accepted to check the value (even Haskell uses testBit), and `toggle` or `flip` seems fine for flipping the bit value